### DISCUSSION TITLE
How to use elevation of raster file as top of model grid or to set up unstructured grid model

### DISCUSSION THREAD

**Discussion by AH-CSM** [modflowpy/flopy#1644] [Q&A]
Hello all,

I am trying to set up my model so that the elevation data within the raster file I am importing can be used as the top of the model grid. I am doing this because I am trying to set my bottom thicknesses from the top down in layering. The current model will not let me do it because the layer bottoms I want for my model are consistently above SOME of my raster data elevation. In the real world, this would be a pinched-off layer but I am not sure how to set this up in a structured or unstructured grid way to implement raster data for my model top. Any advice is greatly appreciated. Thanks!

**Comment by jlarsen-usgs** [comment]
Hi @AH-CSM;

If you have vertex information stored in your `UnstructuredGrid` object (modelgrid) you can use the flopy.utils.Raster class to resample the DEM raster to grid and effectively create an array of model top elevations. Ex.

```python
from flopy.utils import Raster

rstr = Raster.load("my_dem_raster.tif")
top_elevs = rstr.resample_to_grid(
    modelgrid,
    band=rstr.bands[0],
    method="min"  # methods supported are "min", "max", "median", "mean", "nearest", "linear", "cubic"
)
botm_elev = np.zeros((modelgrid.nnodes))
botm_elev = top_elevs - layer_thickness 

# then update the top and botm elevations in your flopy DISU object using the new values
# Note: if you're working with MF-USG and are using the layered option, you may need to reshape your botm_elevations to be consistent 
# with the number of cells per layer
```

There's a notebook showing the Raster resampling process, here: [raster_intersection](https://github.com/modflowpy/flopy/blob/develop/examples/Notebooks/flopy3_raster_intersection.ipynb). 

Some good tutorials can also be found in the python for hydrology repository here: [python-for-hydrology, flopy](https://github.com/DOI-USGS/python-for-hydrology/tree/main/notebooks/part1_flopy/solutions)

**Reply by AH-CSM** [reply]
I have been trying to use the notebook you provided above as an example for a while now (there is very little literature on how to use a lot of flopy in unique scenarios) to build my own but some of the lines of code do not run properly and I have to find ways around them. For example, my Python IDE says

"Exception: Cannot find triangle binary executable"

when I try to run the unstructured grid with a triangle module call like your above example shows. I have tried looking in my platform to update flopy or other modules in anaconda but nothing for triangles or utils comes up in the library directory.

**Reply by jlarsen-usgs** [reply]
@AH-CSM 

Triangle is a external modflow related executable program. A full stack of modflow executables can be downloaded from here: [https://github.com/MODFLOW-USGS/executables](https://github.com/MODFLOW-USGS/executables).

Alternatively, once flopy is installed you can run `get-modflow :flopy` in an anaconda prompt and it will download all of the necessary modflow executables and place them within the anaconda/python path.
